### PR TITLE
[WIP] Add all standard longhand CSS properties to Box as style props

### PR DESCRIPTION
**Pull request by jesstelford**
The goal here is to make access to Polaris design tokens easier than writing raw CSS via an API that aligns with admin builders' expectations, while still exposing close to 100% of CSS's capabilities.

> [!NOTE]
> **This PR is an experiment and unlikely to ship in its current form (or at all).**

## Goals

- [x] Where possible make access to Polaris design tokens _easier_ than writing raw CSS/SASS
    - The logical place for this is React props since we already embrace the HTML-in-JS paradigm.
- [x] Expose as close to 100% of the capability of raw CSS as possible, expanding the [current Polaris `<Box>` style props API](https://polaris.shopify.com/components/layout-and-structure/box#props).
- [x] Provide an API which matches admin builder's expectations of how CSS functions.
- [x] Donut Scope styles to the DOM node being rendered
    - ie; Don't allow styles from a parent node to influence this node, and don't allow this node's styles influence any children. 
- [ ] Remove usage of SASS entirely.
- [x] Strongly type design token values into all style props for tighter IDE autocomplete integration.
    - This shortens the DX loop by cutting out "look it up in the docs site" as a step.
- [x] Reduce the linear relationship between `<Box>` API additions and code complexity / size (each new style prop requires multiple lines of CSS to work) to a constant value by offloading as much work as possible to the web platform (ie; browsers).
- [x] Maintain support for Server Side Rendering.
- [x] Support streaming responses.
- [x] Reduce up-front network requests and code size required for the browser to begin rendering the page (ie; remove massive .css files that must be parsed before anything can be rendered).
- [x] Maintain current build time and run time performance.

## How it's done

**tl;dr**:
- All style props are passed directly to the DOM element's `style` attribute.
- Design token values are automatically converted to their CSS custom property counterparts.
- Advanced (spec compliant) CSS techniques are used to enable responsive style values on the `style` attribute with O(1) `.css` size / complexity increase.

### Attempt 1: Extending the existing API

Our original approach to this problem was extending the existing implementation in `@shopify/polaris`'s `<Box>` from 39 supported CSS properties to _all_ (shortand) CSS properties (approx 400).

This worked, but generated a massive .css file of 1.1mb due to the combined scoping & media query styles. We were able to find some optimisations in that technique, but it was still a significant increase in CSS size, and didn't account for any style modifiers (`:hover`, etc) at all (each of which would multiply the size of the generated output).

### Attempt 2 (this PR): CSS Space Hack

During our research, we discovered the [CSS Space Hack](https://lea.verou.me/blog/2020/10/the-var-space-hack-to-toggle-multiple-values-with-one-custom-property/), and specifically the [CSS Media Vars](https://propjockey.github.io/css-media-vars/) implementation which showed a CSS spec-compliant way to do media queries in an inline `style=` attribute. _(Update: We've since been made aware of [css-hooks](https://css-hooks.com/) which achieves a similar end result via the CSS Space Hack technique)_

This discovery led to a reimplementation of the prototype using CSS Space Hack that solved the CSS Custom Variable scoping problem (by not needing them at all), and so doesn't require any CSS declarations at all in a CSS file (beyond the ~20 lines of CSS Space Hack setup).

Ie; our CSS file went from ~1,136,000 bytes to **just 368 bytes** for all 400 CSS properties and all 5 media queries! Each modifier (`:hover`), etc, only adds 260 bytes.

#### Caveats

There are some drawbacks to the CSS Space Hacks approach:

1. A slight increase in the size of the inline `style=` attribute (_heavily offset by the reduction in the `.css` file_)
2. Inline `style=` declarations have a much higher specificity than classes which may target that DOM node.
    - However, this is actually aligned with one of the stated goals of this prototype: to stop builders from stepping outside the system by applying classnames.
3. No support for pseudo selectors (`::before` / `::after`)
    - Note: Only 400 uses in the admin
5. Possible rendering performance impact within React (to be confirmed)

#### Generating types

A one time build script is used to generate the relevant types by merging our design token types from `@shopify/polaris-tokens` into the types exported by [`csstype`](https://github.com/frenic/csstype).

To support the existing `<Box>` API where `padding` is mapped to `paddingInlineStart`/`paddingInlineEnd`/`paddingBlockStart`/`paddingBlockEnd`, we generate "alias" types as the intersection (_the Set Theory definition of the word_) of the target properties to ensure only valid values are expanded onto the concrete `style` attribute.

For example; `backgroundPosition` might be an alias for setting both `backgroundPositionX` and `backgroundPositionY`:

```
const backgroundPositionX = props.backgroundPositionX ?? props.backgroundPosition;
const backgroundPositionY = props.backgroundPositionY ?? props.backgroundPosition;
```

However, `left` isn't a valid value for `backgroundPositionY`, nor is `top` a valid value for `backgroundPositionX`. So the script generates the type as:

```
  backgroundPosition?: Exclude<
    csstypes.LonghandStyleProps['backgroundPositionX'],
    'left'
    | 'right'
    | 'x-end'
    | 'x-start'
  >;
```

Which is the intersection (_the Set Theory definition of the word_) of `csstypes.LonghandStyleProps['backgroundPositionX']` & `csstypes.LonghandStyleProps['backgroundPositionY']`.

## Reading this PR

[`polaris-react/src/components/Box/generated-style.module.scss`](https://github.com/Shopify/polaris/blob/style-props-space-hack/polaris-react/src/components/Box/generated-style.module.scss)
The core mechanism of the CSS Space Hack technique. Currently generated by a script, but could easily be hand written exactly the same way.

[`polaris-react/src/components/Box/get-style-props.ts`](https://github.com/Shopify/polaris/blob/style-props-space-hack/polaris-react/src/components/Box/get-style-props.ts)
The meat of this PR where props get turned into `style` attributes:
- `paddingInlineStart="20px"` becomes `style="padding-inline-start: 20px"`
- handling responsive values: `paddingInlineStart={{ xs: "20px", lg: "40px" }}` becomes roughly `style="--pad-lg: var(--_lg) 40px; padding-inline-start: var(--pad-lg, 20px)"`
- handling modifiers: `paddingInlineStart={{ _hover: "20px" }}` becomes roughly `style="--pad-hov: var(--__hover) 20px; padding-inline-start: var(--pad-hov)"`
- and combinations of the above

[`polaris-react/src/components/Box/generated-data.ts`](https://github.com/Shopify/polaris/blob/style-props-space-hack/polaris-react/src/components/Box/generated-data.ts)
Strong Typescript types for all style props including design tokens and aliases. Generated based on config in [`polaris-react/scripts/get-style-props/data.mjs`](https://github.com/Shopify/polaris/blob/style-props-space-hack/polaris-react/scripts/get-style-props/data.mjs).

[`polaris-react/src/components/Box/Box.tsx`](https://github.com/Shopify/polaris/blob/style-props-space-hack/polaris-react/src/components/Box/Box.tsx)
The React call site for all of this, plus the concrete implementation of converting design token prop values into CSS custom vars.

### CODE REVIEW

File: polaris-react/src/components/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -129,8 +129,8 @@ export const Checkbox = forwardRef<CheckboxHandles, CheckboxProps>(
     const isChecked = !isIndeterminate && Boolean(checked);
 
     const indeterminateAttributes = isIndeterminate
-      ? {indeterminate: 'true', 'aria-checked': 'mixed' as const}
-      : {'aria-checked': isChecked};
+      ? {indeterminate: 'true'}
+      : {};
```
Comment (jesstelford) [inline polaris-react/src/components/Checkbox/Checkbox.tsx:133]: The accessibility audit was throwing [this error](https://dequeuniversity.com/rules/axe/4.8/aria-conditional-attr?application=axe-puppeteer) here... It's seemingly unrelated to our changes, but I couldn't get CI green without it, so 🤷

File: package.json [hunk 2]
```diff
@@ -43,11 +43,11 @@
...
-    "@babel/core": "^7.20.12",
-    "@babel/node": "^7.20.7",
-    "@babel/preset-env": "^7.16.4",
-    "@babel/preset-react": "^7.16.0",
-    "@babel/preset-typescript": "^7.18.6",
+    "@babel/core": "^7.23.3",
+    "@babel/node": "^7.22.19",
+    "@babel/preset-env": "^7.23.3",
+    "@babel/preset-react": "^7.23.3",
+    "@babel/preset-typescript": "^7.23.3",
```
Comment (jesstelford) [inline package.json:50]: Had to bump versions to get proper support for TS `satisfies`, particularly in `jest` (which uses Babel under the hood to transpile `.tsx` files).

File: polaris-for-vscode/src/server.ts [hunk 3]
```diff
@@ -20,7 +20,7 @@ import {TextDocument} from 'vscode-languageserver-textdocument';
 
 const excludedTokenGroupNames = [] as const;
 
-type ExcludedTokenGroupName = typeof excludedTokenGroupNames[number];
```
Comment (jesstelford) [inline polaris-for-vscode/src/server.ts:23]: Prettier did this.

File: polaris-migrator/src/migrate.ts [hunk 4]
```diff
@@ -1,7 +1,6 @@
 import fs from 'fs';
 import path from 'path';
 
-// @ts-expect-error TS can't resolve the type of this import
 import * as jscodeshift from 'jscodeshift/src/Runner';
```
Comment (jesstelford) [inline polaris-migrator/src/migrate.ts:4]: I guess TS _can_ resolve these types now, so it told me to remove the comment. Which I did, and there's no error 🌮

File: polaris-migrator/src/utilities/sass.ts [hunk 5]
```diff
@@ -489,7 +489,7 @@ export function createSassMigrator(name: string, ruleFn: PolarisMigrator) {
       const {walker, serialiseSuggestion} = args;
 
       return (node: T) => {
-        let oldNode: T;
```
Comment (jesstelford) [inline polaris-migrator/src/utilities/sass.ts:492]: TS complained about `node` being instantiated with a different sub type than `oldNode`, but giving it a concrete type here instead of the generic fixed it.

File: polaris-react/.storybook/RenderPerformanceProfiler/RenderPerformanceProfiler.tsx [hunk 6]
```diff
@@ -3,7 +3,7 @@ import React, {PropsWithChildren} from 'react';
 interface Data {
   id: string;
   kind: string;
-  phase: 'mount' | 'update';
```
Comment (jesstelford) [inline polaris-react/.storybook/RenderPerformanceProfiler/RenderPerformanceProfiler.tsx:6]: This type was outdated; the latest `@types/react-dom` adds `'nested-update'`.

File: polaris-react/src/components/Box/Box.scss [hunk 7]
```diff
@@ -8,113 +8,6 @@
...
-  outline-width: var(--pc-box-outline-width);
-  // stylelint-disable-next-line -- generated by polaris-migrator DO NOT COPY
-  overflow-x: var(--pc-box-overflow-x);
-  // stylelint-disable-next-line -- generated by polaris-migrator DO NOT COPY
-  overflow-y: var(--pc-box-overflow-y);
-  // stylelint-disable-next-line -- generated by polaris-migrator DO NOT COPY
-  width: var(--pc-box-width);
-  -webkit-overflow-scrolling: touch;
-}
-
```
Comment (jesstelford) [inline polaris-react/src/components/Box/Box.scss]: 🎉

File: polaris-react/src/components/Box/Box.tsx [hunk 8]
```diff
@@ -1,346 +1,184 @@
...
-        : undefined,
-      '--pc-box-inset-inline-start': insetInlineStart
-        ? `var(--p-space-${insetInlineStart})`
-        : undefined,
-      '--pc-box-inset-inline-end': insetInlineEnd
-        ? `var(--p-space-${insetInlineEnd})`
-        : undefined,
-      zIndex,
-      opacity,
-    } as React.CSSProperties;
```
Comment (jesstelford) [inline polaris-react/src/components/Box/Box.tsx:321]: 🎉

File: polaris-react/src/components/Box/generated-style.scss [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  --_xl_hover: var(--_xl);
+}
+
+.Box:visited {
+  --__visited: ;
+  --_sm_visited: var(--_sm);
+  --_md_visited: var(--_md);
+  --_lg_visited: var(--_lg);
+  --_xl_visited: var(--_xl);
+}
```
Comment (jesstelford) [inline polaris-react/src/components/Box/generated-style.scss]: See: https://github.com/propjockey/css-media-vars

File: polaris-react/src/components/Button/Button.scss [hunk 10]
```diff
@@ -47,8 +47,12 @@
...
 
   // stylelint-disable selector-max-specificity -- focus styles
-  &:focus:not(.variantPrimary):not(.variantTertiary):not(.variantPlain):not(.variantMonochromePlain):not(.pressed),
-  &:focus-visible:not(.variantPrimary):not(.variantTertiary):not(.variantPlain):not(.variantMonochromePlain):not(.pressed) {
+  &:focus:not(.variantPrimary):not(.variantTertiary):not(.variantPlain):not(
+      .variantMonochromePlain
+    ):not(.pressed),
+  &:focus-visible:not(.variantPrimary):not(.variantTertiary):not(
+      .variantPlain
+    ):not(.variantMonochromePlain):not(.pressed) {
```
Comment (jesstelford) [inline polaris-react/src/components/Button/Button.scss]: Prettier 🤷

File: polaris-react/src/components/Button/Button.scss [hunk 11]
```diff
@@ -963,7 +967,9 @@
   }
 }
 // Overrides for popover active / expanded states on default buttons
-.Button:is([aria-expanded='true']):is([aria-owns]):not(.variantPlain):not(.variantMonochromePlain):not(.variantTertiary):not(.variantPrimary) {
+.Button:is([aria-expanded='true']):is([aria-owns]):not(.variantPlain):not(
+    .variantMonochromePlain
+  ):not(.variantTertiary):not(.variantPrimary) {
```
Comment (jesstelford) [inline polaris-react/src/components/Button/Button.scss]: Prettier 🤷

File: polaris-react/src/components/ChoiceList/ChoiceList.tsx [hunk 12]
```diff
@@ -73,6 +73,9 @@ export function ChoiceList({
   const titleMarkup = title ? (
     <Box
       as="legend"
+      // User agent default styles have a 2px padding around <legend> elements
+      // that we don't want.
+      padding="0"
```
Comment (jesstelford) [inline polaris-react/src/components/ChoiceList/ChoiceList.tsx]: This wasn't an issue before because our styles were inadvertently setting `padding-inline-start: initial` which overrides the user agent specific styles. In the new approach, that isn't the case, so we leverage the styleProp to get it donw.

File: polaris-react/src/components/Divider/Divider.tsx [hunk 13]
```diff
@@ -8,7 +8,7 @@ export interface DividerProps {
    * Divider border color
    * @default 'border-secondary'
    */
-  borderColor?: ColorBorderAlias | 'transparent';
```
Comment (jesstelford) [inline polaris-react/src/components/Divider/Divider.tsx:11]: `transparent` is now a valid token value.

File: polaris-react/src/styles/shared/_accessibility.scss [hunk 14]
```diff
@@ -10,7 +10,7 @@
   // Top position is required to prevent unexpected
   // scrolling with Sheet component
   // https://github.com/Shopify/polaris-react/pull/5208
-  top: 0;
+  top: 0 !important;
```
Comment (jesstelford) [inline polaris-react/src/styles/shared/_accessibility.scss:13]: Don't judge me 😓 This is purely for consistency with the rest of the styles here.

File: polaris-react/src/utilities/css.ts [hunk 15]
```diff
@@ -5,6 +5,10 @@ import {isObject} from './is-object';
 
 type Falsy = boolean | undefined | null | 0;
 
+export type ResponsivePropObject<T> = {
+  [Breakpoint in BreakpointsAlias]?: T;
+};
+
 type ResponsivePropConfig<T = string> = {
   [Breakpoint in BreakpointsAlias]?: T;
 };
```
Comment (jesstelford) [inline polaris-react/src/utilities/css.ts:14]: This type will go away once we replace all the `getResponsiveProps` calls throughout the codebase with usages of `<Box>`

File: polaris-tokens/package.json [hunk 16]
```diff
@@ -56,6 +56,6 @@
     "dist"
   ],
   "dependencies": {
-    "deepmerge": "^4.3.1"
+    "deepmerge-ts": "^5.1.0"
```
Comment (jesstelford) [inline polaris-tokens/package.json:59]: Same functionality, just typed properly.

File: polaris-tokens/src/themes/base/height.ts [hunk 17]
```diff
@@ -93,3 +97,22 @@ export const height: {
     value: size[3200],
   },
 };
+
+const heightStyleProps = [
+  // Logical properties
+  // TODO: Reinstate these once our token scales are usable
+  // 'blockSize',
+  // 'minBlockSize',
+  // 'maxBlockSize',
```
Comment (jesstelford) [inline polaris-tokens/src/themes/base/height.ts:106]: Our size scale stops at `148px` which is pretty much useless for setting width/height on elements. Not to mention all the usages of `100%` just within our own codebase!

These properties _should_ be tokenized, but until we fix the size scale, it'll have to omitted.

File: polaris.shopify.com/package.json [hunk 18]
```diff
@@ -85,13 +92,23 @@
...
+    "tmp": "^0.2.1",
     "ts-node": "^10.7.0",
-    "typescript": "^4.9.3"
+    "typescript": "^4.9.3",
+    "yargs": "^17.7.1"
+  },
+  "resolutions": {
+    "@types/react": "18.2.0",
+    "@types/react-dom": "18.2.0",
+    "playroom": "0.28.0"
```
Comment (jesstelford) [inline polaris.shopify.com/package.json:117]: Have to lock down the version since a new one got released which is conflicting with our patch to it.

File: polaris.shopify.com/src/components/Markdown/next-mdx-importer/serialize.ts [hunk 19]
```diff
@@ -1,7 +1,7 @@
 import type {Plugin} from 'unified';
-import type {Literal} from 'unist';
+import type {Literal, Node} from 'unist';
 import {VFile} from 'vfile';
-import {visit, type Node} from 'unist-util-visit';
```
Comment (jesstelford) [inline polaris.shopify.com/src/components/Markdown/next-mdx-importer/serialize.ts:4]: Honestly nfi why this was broken, but it appears the type simply moved 👍

File: polaris.shopify.com/src/styles/variables.scss [hunk 20]
```diff
@@ -4,7 +4,6 @@ $breakpointMobile: 500px;
 $breakpointTablet: 768px;
 $breakpointDesktopSmall: 1040px;
 $breakpointDesktop: 1400px;
-$breakpointDesktopLarge: 1600px;
```
Comment (jesstelford) [inline polaris.shopify.com/src/styles/variables.scss:7]: Weren't using this breakpoint anywhere, so I removed it.

**Comment by jesstelford** [comment]
Note there is a newer branch with some refactored logic: [`style-props-space-hack-refactor`](https://github.com/Shopify/polaris/tree/style-props-space-hack-refactor).
